crates/ruff/src/rules/ruff/rules/unnecessary_iterable_allocation_for_first_element.rs

```diff
@@ -175,47 +214,98 @@ fn iterable_name<'a>(expr: &'a Expr, model: &SemanticModel) -> Option<&'a str> {
                 return None;
             }
 
+            let [arg] = args.as_slice() else {
+                return None;
+            };
+
             if !model.is_builtin(id.as_str()) {
                 return None;
             }
 
-            match args.first() {
-                Some(Expr::Name(ast::ExprName { id: arg_name, .. })) => Some(arg_name.as_str()),
-                Some(Expr::GeneratorExp(ast::ExprGeneratorExp {
+            match arg {
+                Expr::GeneratorExp(ast::ExprGeneratorExp {
+                    elt, generators, ..
+                }) => match_simple_comprehension(elt, generators)
+                    .map(|range| IterationTarget {
+                        range,
+                        iterable: false,
+                    })
+                    .or_else(|| {
+                        Some(IterationTarget {
+                            range: arg.range(),
+                            iterable: true,
+                        })
+                    }),
+                Expr::ListComp(ast::ExprListComp {
                     elt, generators, ..
-                })) => generator_iterable(elt, generators),
-                _ => None,
+                }) => match_simple_comprehension(elt, generators)
+                    .map(|range| IterationTarget {
+                        range,
+                        iterable: false,
+                    })
+                    .or_else(|| {
+                        Some(IterationTarget {
+                            range: arg
+                                .range()
+                                // Remove the `[`
+                                .add_start(TextSize::from(1))
+                                // Remove the `]`
+                                .sub_end(TextSize::from(1)),
+                            iterable: true,
+                        })
+                    }),
+                _ => Some(IterationTarget {
+                    range: arg.range(),
+                    iterable: false,
+                }),
             }
         }
         Expr::ListComp(ast::ExprListComp {
             elt, generators, ..
-        }) => generator_iterable(elt, generators),
+        }) => match_simple_comprehension(elt, generators)
+            .map(|range| IterationTarget {
+                range,
+                iterable: false,
+            })
+            .or_else(|| {
+                Some(IterationTarget {
+                    range: expr
+                        .range()
+                        // Remove the `[`
+                        .add_start(TextSize::from(1))
+                        // Remove the `]`
+                        .sub_end(TextSize::from(1)),
+                    iterable: true,
+                })
+            }),
         _ => None,
```

Comment:
Nit: This seems to be the only branch that returns `None`. I then prefer to write:

```rust
let result = match arg {
	...
	// Returns no option
	Expr::ListComp(...) => IterationTarget {... },
	
	_ => return None;
};

Some(result)
```

It safes you from having to write that many `Some`'s